server/service/apple_mdm.go

```diff
@@ -3438,6 +3471,249 @@ func ReconcileAppleProfiles(
+			for _, hostUUID := range target.hostUUIDs {
+				err := ds.UpdateOrDeleteHostMDMAppleProfile(ctx, &fleet.HostMDMAppleProfile{
+					CommandUUID: target.cmdUUID,
+					HostUUID:    hostUUID,
+					Status:      &fleet.MDMDeliveryFailed,
+					Detail: "NDES SCEP Proxy is not configured. " +
+						"Please configure in Settings > Integrations > Mobile Device Management > Simple Certificate Enrollment Protocol.",
+					OperationType: fleet.MDMOperationTypeInstall,
+				})
+				if err != nil {
+					return false, err
+				}
+			}
+		}
+		return appConfig.Integrations.NDESSCEPProxy.Valid, nil
+	}
+
+	// Copy of NDES SCEP config which will contain unencrypted password, if needed
+	var ndesConfig *fleet.NDESSCEPProxyIntegration
+
+	var addedTargets map[string]*cmdTarget
+	for profUUID, target := range targets {
+		contents, ok := profileContents[profUUID]
+		if !ok {
+			// This should never happen
+			continue
+		}
+
+		// Check if Fleet variables are present.
+		contentsStr := string(contents)
+		fleetVars := findFleetVariables(contentsStr)
+		if len(fleetVars) == 0 {
+			continue
+		}
+
+		// Do common validation that applies to all hosts in the target
+		valid := true
+		for fleetVar := range fleetVars {
+			switch fleetVar {
+			case FleetVarNDESSCEPChallenge, FleetVarNDESSCEPProxyURL:
+				configured, err := isNDESSCEPConfigured(profUUID, target)
+				if err != nil {
+					return ctxerr.Wrap(ctx, err, "checking NDES SCEP configuration")
+				}
+				if !configured {
+					valid = false
+					break
+				}
+			case FleetVarHostEndUserEmailIDP:
+				// No extra validation needed for this variable
+			default:
+				// Error out if we find an unknown variable
+				for _, hostUUID := range target.hostUUIDs {
+					err := ds.UpdateOrDeleteHostMDMAppleProfile(ctx, &fleet.HostMDMAppleProfile{
+						CommandUUID: target.cmdUUID,
+						HostUUID:    hostUUID,
+						Status:      &fleet.MDMDeliveryFailed,
+						Detail: fmt.Sprintf("Unknown Fleet variable $FLEET_VAR_%s found in profile. Please update or remove.",
+							fleetVar),
+						OperationType: fleet.MDMOperationTypeInstall,
+					})
+					if err != nil {
+						return ctxerr.Wrap(ctx, err, "updating host MDM Apple profile for unknown variable")
+					}
+				}
+				valid = false
+				break
+			}
+		}
+		if !valid {
+			// We marked the profile as failed, so we will not do any additional processing on it
+			delete(targets, profUUID)
+			continue
+		}
+
+		// Currently, all supported Fleet variables are unique per host, so we split the profile into multiple profiles.
+		// We generate a new temporary profileUUID which is currently only used to install the profile.
+		// The profileUUID in host_mdm_apple_profiles is still the original profileUUID.
+		if addedTargets == nil {
+			addedTargets = make(map[string]*cmdTarget, 1)
+		}
+		for _, hostUUID := range target.hostUUIDs {
+			newProfUUID := uuid.NewString()
+			hostContents := contentsStr
+
+			failed := false
+			for fleetVar := range fleetVars {
```

Comment:
can any of the variables in here contain invalid XML characters that need to be escaped? (I see we do escaping of the URL query parameters in a particular case)

Reply:
Yes, we can escape all the variables just to be safe.